Update save.py

nikochiko opened the pull request:
Fix #34348 .
Notes:
- Documentation needs to be changed (in multiple places) after final changes in code.
- Changed code for deciding whether to save file as h5 or tf.
- Removed the unncessary _HDF5_EXTENSIONS list. Will have to make sure it wasn't used elsewhere.
- Added 4 new ValueError raises.

nikochiko committed:
Update array_ops.py

Update documentation, formatting and fix typos for `tf.broadcast_dynamic_shape`, `tf.broadcast_static_shape`, `tf.boolean_mask`

nikochiko committed:
Update docstrings

Updated docstrings for `tf.convert_to_tensor` and `tf.edit_distance`.
`tf.convert_to_tensor`:  Put example in "For example:" section and switch to carets from backticks.
`tf.edit_distance`: Updated documentatoin, fixed example.

nikochiko committed:
Update save.py

Fix https://github.com/tensorflow/tensorflow/issues/34348 .
Notes:
- Documentation needs to be changed (in multiple places) after final changes in code.
- Changed code for deciding whether to save file as h5 or tf.
- Removed the unncessary _HDF5_EXTENSIONS list. Will have to make sure it wasn't used elsewhere.
- Added 4 new ValueError raises.

nikochiko committed:
Revert "Update docstrings"

This reverts commit 2a6efd2e668f8418bdf1c60e8218791559724dc4.

nikochiko committed:
Revert "Update array_ops.py"

This reverts commit 4c9ee36f03d9b01b4d8598905aa26bbf81b380b4.

k-w-w commented on tensorflow/python/keras/saving/save.py:
```diff
@@ -92,12 +90,42 @@ def save_model(model,
-  if (save_format == 'h5' or
-      (h5py is not None and isinstance(filepath, h5py.File)) or
-      os.path.splitext(filepath)[1] in _HDF5_EXTENSIONS):
+  if type(filepath) != str and not isinstance(filepath, h5py.File):
```
Perhaps refactor out this part of the code to a separate function that is called by both `save_weights()` and `save()`. 

Also, it would be useful to have unit tests to make sure that the save format returned by the function is correct

k-w-w reviewed (commented):
This looks good! Thanks for the CL.

nikochiko committed:
Added new function process_save_format

- Added new function `validate_save_format` as requested by @k-w-w inside `network.py`.
- Using `validate_save_format` for validating save_format in `save.save_model` and `network.save_weights`

Although, the a few updates will have to be made in `save_weights` because
- `validate_save_format` is designed to work with path as well as h5py.File objects. This works with `save.save_model` but not with `network.save_weights` which accepts only String as the path.
- Does it make sense to add functionality to save_weights to save it to a h5py.File object?

nikochiko commented on tensorflow/python/keras/saving/save.py:
```diff
@@ -23,6 +23,7 @@
 import six
 
 from tensorflow.python import tf2
+from tensorflow.python.keras.engine.network import _is_hdf5_filepath
```
Done.

nikochiko commented:
@k-w-w , Please see the new changes:
- Added new function `validate_save_format` as requested by @k-w-w inside `network.py`.
- Using `validate_save_format` for validating save_format in `save.save_model` and `network.save_weights`

Although, the a few updates will have to be made in `save_weights` because
- `validate_save_format` is designed to work with path as well as h5py.File objects. This works with `save.save_model` but not with `network.save_weights` which accepts only String as the path.
- Does it make sense to add functionality to save_weights to save it to a h5py.File object?

nikochiko committed:
Merge branch 'master' into fix-save-model

rthadur commented:
@k-w-w gentle ping to review new changes ?

k-w-w reviewed (dismissed):
LGTM Thanks!

gbaned commented:
@nikochiko Could you please check failed build errors? Thanks!

nikochiko commented:
@gbaned Looking into it now. Sorry for the delay,

nikochiko committed:
Update save.py

nikochiko committed:
Update network.py

nikochiko commented:
@gbaned Fixed.

nikochiko committed:
Fix sanity

nikochiko commented:
@gbaned @k-w-w Should I add tests for the new function in the same PR or create another?
Also, can you point me to where the tests should be written.

rthadur commented:
@nikochiko please add the test cases in `tensorflow/python/keras/engine/network_test.py`

nikochiko committed:
Add tests

nikochiko committed:
Fix typo

nikochiko committed:
Fix typo

nikochiko commented:
@rthadur I have added the tests. :)

k-w-w reviewed (dismissed):
Thanks again!

rthadur commented:
@nikochiko can you please fix sanity checks.

nikochiko committed:
Fix spacing

nikochiko commented:
@rthadur done!

rthadur commented:
@nikochiko can you please resolve the conflicts as well ?

nikochiko committed:
Merge branch 'master' into fix-save-model

nikochiko commented:
@rthadur :+1: Done!

k-w-w commented:
Oops, another change is needed to push this. 

Can you modify the line `from tensorflow.python.keras.engine import network` in `save.py` to use the LazyLoader? An example is here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/saving/saved_model/save_impl.py#L45

k-w-w reviewed (changes requested):
(see comment above)

nikochiko committed:
Use LazyLoader to import network in save.py

nikochiko commented:
Thanks for the example! 
I've added the changes :+1:

k-w-w reviewed (dismissed):
Thanks again!

nikochiko committed:
Fix typo

nikochiko commented:
Oops I had missed a trivial typo 😅 
Sorry for requesting review again.

k-w-w reviewed (dismissed):
No problem!

rthadur commented:
@nikochiko can you please fix build failures ?

nikochiko committed:
Fix typo

nikochiko commented:
@rthadur Done!

gbaned commented:
@nikochiko Could you please address Ubuntu Sanity errors? Thanks!

nikochiko committed:
Fix linting

nikochiko commented:
@gbaned Done :+1:

rthadur commented:
> `Traceback (most recent call last):
  File "/build/work//google3/runfiles/google3/third_party/py/absl/third_party/unittest3_backport/case.py", line 37, in testPartExecutor
    yield
  File "/build/work//google3/runfiles/google3/third_party/py/absl/third_party/unittest3_backport/case.py", line 162, in run
    testMethod()
  File "/build/work//google3/runfiles/google3/learning/brain/contrib/distribute/python/keras_example_test_tpu.py", line 39, in test_fit_eval_predict
    callbacks=[tb_callback, checkpoint_cb])
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/keras/engine/training_v1.py", line 792, in fit
    use_multiprocessing=use_multiprocessing)
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/keras/engine/training_distributed.py", line 670, in fit
    validation_freq=validation_freq)
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/keras/engine/training_distributed.py", line 280, in experimental_tpu_fit_loop
    callbacks.on_epoch_end(epoch, epoch_logs)
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/keras/callbacks.py", line 302, in on_epoch_end
    callback.on_epoch_end(epoch, logs)
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/keras/callbacks.py", line 1025, in on_epoch_end
    self._save_model(epoch=epoch, logs=logs)
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/keras/callbacks.py", line 1073, in _save_model
    self.model.save(filepath, overwrite=True)
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/keras/engine/network.py", line 1005, in save
    signatures, options)
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/keras/saving/save.py", line 117, in save_model
    save_format = network.validate_save_format(filepath, save_format)
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/util/lazy_loader.py", line 62, in __getattr__
    module = self._load()
  File "/build/work//google3/runfiles/google3/third_party/tensorflow/python/util/lazy_loader.py", line 45, in _load
    module = importlib.import_module(self.__name__)
  File "<embedded stdlib>/importlib/__init__.py", line 37, in import_module
    __import__(name)
ImportError: No module named python.keras.engine.network`

rthadur commented:
@nikochiko sorry for the delay , this PR has been failing for above internal error, can you please fix this ?

cc @k-w-w @goldiegadde

nikochiko commented:
@rthadur @k-w-w I am unsure as to what is causing the error. I will try to troubleshoot it on my own but please let me know if it is a known issue.
Should the local name be `network_lib` instead of `network`?

k-w-w commented on tensorflow/python/keras/saving/save.py:
```diff
@@ -18,18 +18,21 @@
 
+network = LazyLoader(
+    'network', globals(),
+    'tensorflow.python.keras.engine.network')
```
There's an internal bug, so this has to use double quotes:
```
# pylint:disable=g-inconsistent-quotes
network = LazyLoader(
    "network", globals(),
    "tensorflow.python.keras.engine.network")
# pylint:enable=g-inconsistent-quotes
```

nikochiko replied:
Thanks!

nikochiko replied:
Done! :+1:

nikochiko committed:
Single quotes -> double quotes for LazyLoader

nikochiko committed:
Make pylint comments consistent
